crates/parser/src/queue_stack.rs

```diff
@@ -0,0 +1,209 @@
+/// This module implements a Stack, which is useful for implementing a parser
+/// with variable lookahead, as it would allow to pop elements which are below
+/// the top-element, and maintain a top counter which would be in charge of
+/// moving these elements once shifted.
+use std::{ops::Deref, ptr};
+
+/// This container implements a stack and a queue in a single vector:
+///   - stack: buf[..top]
+///   - queue: buf[top + gap..]
+///
+/// Elements are pushed on the stack by queueing elements in the queue, and
+/// shifting them to the stack.
+///
+/// Elements are removed from the stack with the ability to replace the replaced
+/// elements by a single one, which is added in front of all elements which were
+/// queued previously. The replacing elements would be the next element to be
+/// shifted.
+///
+/// In the context of an LR parser, the stack contains shifted elements, and the
+/// queue contains the lookahead. If the lexer is completely independent of the
+/// parser, all tokens could be queued before starting the parser.
+pub(crate) struct QueueStack<T> {
```

Comment:
Please make this `pub`.

This makes the feature visible everywhere the enclosing module is visible.

Reply:
Any reason to do that, my understanding what that crate have to maintain their external API, while having the ability to have private internals which are public within the crate?